### ISSUE TITLE
YSQL: error: Operation failed ... Value write after transaction start

### ISSUE THREAD

**Issue by naveennazimudeen** [yugabyte/yugabyte-db#4387]
Jira Link: [DB-1402](https://yugabyte.atlassian.net/browse/DB-1402)
The Use Case has two steps:
**Transaction 1** - Select query to check for an existing record with an id
**Transaction 2** - If exists, delete and create a new record

This use case was executed simultaneously from an app open in two browser tabs. Both requests would look for the same record. One request succeeds, and the second parallel request caused yugabyte to throw an error:
```Value write after transaction start``` or ```Error during commit: Operation expired: Transaction expired or aborted by a conflict: 40001```

I tried the same scenario on postgres. Postgres does not throw an error. The second parallel request detects that record does not exist in step 1 of the use case. 

When I repeat this test with Yugabyte several times, there are times when the second parallel request to Yugabyte also sees that the record has been deleted in step 1. Roughly I would say it is a 50-50 split with Yugabyte. With Postgres it always stopped at step 1 for the second parallel request.

**Comment by kutayzorlu** [comment]
the same on my side.

**Comment by kutayzorlu** [comment]
solved with  : 

:

<property name="hibernate.connection.isolation">2</property>

Where

    1: READ UNCOMMITTED
    2: READ COMMITTED  ->  

do not use  Serialized :  8  ! .. locking fully ..!

**Comment by m-iancu** [comment]
@naveennazimudeen A quick follow-up to make sure I understand the scenario:
> Transaction 1 - Select query to check for an existing record with an id

So this is a standalone txn with just one single-key select.

> Transaction 2 - If exists, delete and create a new record

And depending on the result of txn1 select, we either do nothing or run txn 2 which is a delete+insert (into the same key).

Then the full run is running two of these in parallel (so 4 transactions in total).
Is that correct?

Also, what isolation level were you using? One notable difference between Yugabyte and Postgres is that the default in Yugabyte is stricter than Postgres (i.e. `REPEATABLE READ` vs `READ COMMITTED`).

**Comment by def-** [comment]
This also happened during upgrade testing when verifying objects, from 2.13.2.0-b135 to 2.17.1.0-b190: https://phabricator.dev.yugabyte.com/D20539
```
Operation failed. Try again: Flush: Value write after transaction start: { physical: 1670647203740256 } >= { physical: 1670647203325285 }: kConflict: . Errors from tablet servers: [Operation failed. Try again (yb/docdb/conflict_resolution.cc:823): Flush: Value write after transaction start: { physical: 1670647203740256 } >= { physical: 1670647203325285 }: kConflict (transaction error 3)]
```

**Comment by def-** [comment]
This also leads to errors during upgrade_ysql, which I can't ignore so easily:
```
2022-12-12 14:44:23,878 ssh_utils.py:272 INFO testupgrade-aws-rf3-upgrade-2.12.4.2_1 Running cmd: /home/yugabyte/master/bin/yb-admin --timeout_ms 120000 --master_addresses 10.9.196.217:7100,10.9.105.50:7100,10.9.134.11:7100 upgrade_ysql
2022-12-12 14:45:07,248 ssh_utils.py:300 INFO testupgrade-aws-rf3-upgrade-2.12.4.2_1 CMD=/home/yugabyte/master/bin/yb-admin --timeout_ms 120000 --master_addresses 10.9.196.217:7100,10.9.105.50:7100,10.9.134.11:7100 upgrade_ysql, RETCODE=1, OUT=, ERR=Error running upgrade_ysql: Network error (yb/yql/pgwrapper/libpq_utils.cc:271): Unable to upgrade YSQL cluster: Failed to apply migration 'V20__11928__stat_progress_copy.sql' to a database template1: Execute failed: 7, message: ERROR:  Operation failed. Try again: Flush: Value write after transaction start: { physical: 1670856307186283 } >= { physical: 1670856307057724 }: kConflict: . Errors from tablet servers: [Operation failed. Try again (yb/docdb/conflict_resolution.cc:795): Flush: Value write after transaction start: { physical: 1670856307186283 } >= { physical: 1670856307057724 }: kConflict (transaction error 3)] (pgsql error 40001) (aux msg ERROR:  Operation failed. Try again: Flush: Value write after transaction start: { physical: 1670856307186283 } >= { physical: 1670856307057724 }: kConflict: . Errors from tablet servers: [Operation failed. Try again (yb/docdb/conflict_resolution.cc:795): Flush: Value write after transaction start: { physical: 1670856307186283 } >= { physical: 1670856307057724 }: kConflict (transaction error 3)])
2022-12-12 14:45:07,249 test_universe_base.py:536 ERROR testupgrade-aws-rf3-upgrade-2.12.4.2_1 '/home/yugabyte/master/bin/yb-admin --timeout_ms 120000 --master_addresses 10.9.196.217:7100,10.9.105.50:7100,10.9.134.11:7100 upgrade_ysql' returned error code 1.
2022-12-12 14:45:07,249 test_base.py:165 ERROR testupgrade-aws-rf3-upgrade-2.12.4.2_1 ITEST FAILED testupgrade-aws-rf3-upgrade-2.12.4.2_1 : RuntimeError(RuntimeError("'/home/yugabyte/master/bin/yb-admin --timeout_ms 120000 --master_addresses 10.9.196.217:7100,10.9.105.50:7100,10.9.134.11:7100 upgrade_ysql' returned error code 1."))
2022-12-12 14:45:07,250 test_base.py:166 INFO testupgrade-aws-rf3-upgrade-2.12.4.2_1 Traceback (most recent call last):
  File "/var/lib/jenkins/code/internal-services/itest/src/universe_tests/test_universe_base.py", line 530, in run_yb_admin
    out_str, err_str = run_and_check_ssh_cmd(client, cmd, timeout=timeout_sec)
  File "/var/lib/jenkins/code/internal-services/itest/src/ssh_utils.py", line 307, in run_and_check_ssh_cmd
    node_cmd, ret_code))
RuntimeError: '/home/yugabyte/master/bin/yb-admin --timeout_ms 120000 --master_addresses 10.9.196.217:7100,10.9.105.50:7100,10.9.134.11:7100 upgrade_ysql' returned error code 1.
```

**Comment by BrettHoyer** [comment]
I hit this last night when deleting and immediately recreating a table on a multi-region YBM deployment

**Comment by sushantrmishra** [comment]
The original reported problem can be resolved by using wait_queue which is slated to be released in 2.19.2 release.
